Adding a shared device test suite (#695)
XanaduAI/pennylane: .github/CHANGELOG.md

<h3>New features since last release</h3>

* Adds a device test suite, located at `pennylane/plugins/tests`, which can be used 
  to run generic tests on core or external devices calling 
  

>   >>> pytest pennylane/plugins/tests --device default.qubit --shots 1234 --analytic False                                                                                                                                                                                                                                                                     >

Author: ```suggestion
  >>> pytest pennylane/plugins/tests --device default.qubit --shots 1234 --analytic False                                                                                                                                                                                                                                                                    
```

  
  The command line arguments are optional.
   
  * If `--device` is not given, the tests are run on the core devices that ship with PennyLane. 
  

>   * If `--shots` is not given, a default of 50000 is used

Author: ```suggestion
  * If `--shots` is not given, a default of 50000 is used.
```

>   * If `--analytic` is not given, the device's default is used

Author: ```suggestion
  * If `--analytic` is not given, the device's default is used.
```

>   If the test is run on external devices, the device and its dependencies must be 

Author: ```suggestion
  If the tests are run on external devices, the device and its dependencies must be 
```

<h3>New features since last release</h3>

> * Adds a device test suite, located at `pennylane/plugins/tests`, which can be used 

Reviewer: @mariaschuld ever since you mentioned it, I definitely thing we should rename `pennylane.plugins` to `pennylane.devices`

  to run generic tests on core or external devices calling 
  

>   >>> pytest pennylane/plugins/tests --device default.qubit --shots 1234 --analytic False                                                                                                                                                                                                                                                                    

Reviewer: 🎉

Reviewer: I did some research on this, and it looks like this is the exact solution we eventually want: https://stackoverflow.com/questions/39091443/running-py-test-tests-from-another-pip-installed-package-against-local-code:

> I am building a package that provides (1) a fixed interface with (2) pluggable drivers. The desired driver will be selectable by the user of the package. My package will include at least one driver with it, but I would like other developers to be able to implement and validate drivers that conform to the package's interface. As such, I would like these developers to be able to run my tests against their drivers.

---

As a short term solution, we could:

* Add the function `qml.get_device_tests()` that returns the test location on the users system (similar to [`np.get_include()`](https://numpy.org/devdocs/reference/generated/numpy.get_include.html).

* Provide a convenience CLI script alongside PennyLane, that is simply a wrapper around `pytest qml.get_device_tests()`.

Author: I added this to the list of next milestones!

  
  The command line arguments are optional.
   
  * If `--device` is not given, the tests are run on the core devices that ship with PennyLane. 
  

>   * If `--shots` is not given, a default of 50000 is used.

Reviewer 2: Would 50000 be a bit too many for default? :thinking: We have a 1000 as the default for the device creation. Or could it make sense to have a separate default number of shots for analytic and non-analytic cases?

Reviewer: Oh, this is a good point @antalszava --- Qiskit has a maximum of 8192 shots as well!

Author: Cool, let me set 10,000

Reviewer 2: @mariaschuld just double-checking: would it be 10,000 or 1,000 here?

  
  * If `--analytic` is not given, the device's default is used.
  

>   Other arguments of the device, such as `strawberryfields.fock`'s compulsory `cutoff_dim`, 

Reviewer 2: (Minor suggestion): just to have a qubit model example for this CHANGELOG entry (as CV would be upcoming)
```suggestion
  Other arguments of the device, such as `qiskit.aer`'s compulsory `backend_options`, 
```

>   can be defined in the `config.toml` file containing custom PennyLane configurations.

Reviewer: :+1:

<h4>New and improved simulators</h4>

* Added a new device, `default.qubit.tf`, a pure-state qubit simulator written using TensorFlow.

> 

Reviewer: ```suggestion
```
